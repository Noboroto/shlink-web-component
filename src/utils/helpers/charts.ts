import type { ActiveElement, ChartEvent, ChartType, TooltipItem } from 'chart.js';
import { prettify } from './numbers';

export const pointerOnHover = ({ native }: ChartEvent, [firstElement]: ActiveElement[]) => {
  if (!native?.target) {
    return;
  }

  const canvas = native.target as HTMLCanvasElement;

  canvas.style.cursor = firstElement ? 'pointer' : 'default';
};

export const renderChartLabel = ({ dataset, raw }: TooltipItem<ChartType>) => `${dataset.label}: ${prettify(`${raw}`)}`;

export const renderPieChartLabel = ({ label, raw }: TooltipItem<ChartType>) => `${label}: ${prettify(`${raw}`)}`;
