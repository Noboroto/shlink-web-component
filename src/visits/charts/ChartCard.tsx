import type { FC, PropsWithChildren, ReactNode } from 'react';
import { Card, CardBody, CardFooter, CardHeader } from 'reactstrap';
import './ChartCard.scss';

type ChartCardProps = PropsWithChildren<{
  title: Function | string;
  footer?: ReactNode;
}>;

export const ChartCard: FC<ChartCardProps> = ({ title, footer, children }) => (
  <Card role="document">
    <CardHeader className="chart-card__header">{typeof title === 'function' ? title() : title}</CardHeader>
    <CardBody>{children}</CardBody>
    {footer && <CardFooter className="chart-card__footer--sticky">{footer}</CardFooter>}
  </Card>
);
