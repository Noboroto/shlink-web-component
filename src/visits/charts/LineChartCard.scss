@import '@shlinkio/shlink-frontend-kit/base';

.line-chart-card__body canvas {
  height: 300px !important;

  @media (min-width: $mdMin) {
    height: 400px !important;
  }
}
