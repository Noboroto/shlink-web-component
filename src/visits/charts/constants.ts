import { isDarkThemeEnabled } from '@shlinkio/shlink-frontend-kit';
import type { CSSProperties } from 'react';
import type { TooltipProps } from 'recharts';

export const CHART_TOOLTIP_STYLES: CSSProperties = {
  color: 'white',
  background: 'rgb(0 0 0 / .9)',
  border: 'none',
  borderRadius: '5px',
} as const;

const CHART_TOOLTIP_LABEL_STYLES: CSSProperties = {
  marginBottom: '5px',
  fontWeight: 'bold',
} as const;

export const CHART_TOOLTIP_COMMON_PROPS: Pick<
  TooltipProps<any, any>,
  'contentStyle' | 'labelStyle' | 'itemStyle'
> = {
  contentStyle: CHART_TOOLTIP_STYLES,
  labelStyle: CHART_TOOLTIP_LABEL_STYLES,
  itemStyle: { padding: 0 },
} as const;

const COLORS = [
  '#97BBCD',
  '#F7464A',
  '#46BFBD',
  '#FDB45C',
  '#949FB1',
  '#57A773',
  '#414066',
  '#08B2E3',
  '#B6C454',
  '#DCDCDC',
  '#463730',
] as const;

export const chartColorForIndex = (index: number) =>
  COLORS[index % COLORS.length];

export const prevColor = () => (isDarkThemeEnabled() ? '#46e587' : '#1DAA58');

export const prevColorAlpha = () =>
  isDarkThemeEnabled() ? 'rgba(70, 229, 135, 0.4)' : 'rgba(29, 170, 88, 0.4)';
