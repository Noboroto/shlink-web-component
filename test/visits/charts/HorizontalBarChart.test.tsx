import type { HorizontalBarChartProps } from '../../../src/visits/charts/HorizontalBarChart';
import { HorizontalBarChart } from '../../../src/visits/charts/HorizontalBarChart';
import { checkAccessibility } from '../../__helpers__/accessibility';
import { setUpCanvas } from '../../__helpers__/setUpTest';

describe('<HorizontalBarChart />', () => {
  const setUp = (props: Omit<HorizontalBarChartProps, 'label'>) => setUpCanvas(
    <HorizontalBarChart label="The chart" {...props} />,
  );

  it('passes a11y checks', () => checkAccessibility(setUp({ stats: {} })));

  it.each([
    [{ foo: 123, bar: 456 }, undefined],
    [{ one: 999, two: 131313 }, { one: 30, two: 100 }],
    [{ one: 999, two: 131313, max: 3 }, { one: 30, two: 100 }],
  ])('renders chart with expected canvas', (stats, highlightedStats) => {
    const { events } = setUp({ stats, highlightedStats });

    expect(events).toBeTruthy();
    expect(events).toMatchSnapshot();
  });
});
