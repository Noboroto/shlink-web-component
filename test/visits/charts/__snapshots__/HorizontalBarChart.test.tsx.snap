// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<HorizontalBarChart /> > renders chart with expected canvas 1`] = `
<div>
  <div
    class="recharts-wrapper"
    role="region"
    style="position: relative; cursor: default; width: 800px; height: 400px;"
  >
    <svg
      class="recharts-surface"
      height="400"
      style="width: 100%; height: 100%;"
      viewBox="0 0 800 400"
      width="800"
    >
      <title />
      <desc />
      <defs>
        <clippath
          id="recharts3-clip"
        >
          <rect
            height="360"
            width="690"
            x="105"
            y="5"
          />
        </clippath>
      </defs>
      <g
        class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
      >
        <line
          class="recharts-cartesian-axis-line"
          fill="none"
          height="30"
          orientation="bottom"
          stroke="#666"
          width="690"
          x="105"
          x1="105"
          x2="795"
          y="365"
          y1="365"
          y2="365"
        />
        <g
          class="recharts-cartesian-axis-ticks"
        >
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="105"
              x2="105"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="105"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="105"
              >
                0
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="277.5"
              x2="277.5"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="277.5"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="277.5"
              >
                150
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="450"
              x2="450"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="450"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="450"
              >
                300
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="622.5"
              x2="622.5"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="622.5"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="622.5"
              >
                450
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="795"
              x2="795"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="795"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="795"
              >
                600
              </tspan>
            </text>
          </g>
        </g>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
      >
        <line
          class="recharts-cartesian-axis-line"
          fill="none"
          height="360"
          orientation="left"
          stroke="#666"
          style="font-size: .8rem;"
          width="100"
          x="5"
          x1="105"
          x2="105"
          y="5"
          y1="5"
          y2="365"
        />
        <g
          class="recharts-cartesian-axis-ticks"
        >
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              style="font-size: .8rem;"
              width="100"
              x="5"
              x1="99"
              x2="105"
              y="5"
              y1="95"
              y2="95"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              style="font-size: .8rem;"
              text-anchor="end"
              width="100"
              x="97"
              y="95"
            >
              <tspan
                dy="0.355em"
                x="97"
              >
                foo
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              style="font-size: .8rem;"
              width="100"
              x="5"
              x1="99"
              x2="105"
              y="5"
              y1="275"
              y2="275"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              style="font-size: .8rem;"
              text-anchor="end"
              width="100"
              x="97"
              y="275"
            >
              <tspan
                dy="0.355em"
                x="97"
              >
                bar
              </tspan>
            </text>
          </g>
        </g>
      </g>
      <g
        class="recharts-cartesian-grid"
      >
        <g
          class="recharts-cartesian-grid-horizontal"
        >
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="95"
            y2="95"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="275"
            y2="275"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="5"
            y2="5"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="365"
            y2="365"
          />
        </g>
        <g
          class="recharts-cartesian-grid-vertical"
        >
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="105"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="277.5"
            x2="277.5"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="450"
            x2="450"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="622.5"
            x2="622.5"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="795"
            x2="795"
            y="5"
            y1="5"
            y2="365"
          />
        </g>
      </g>
      <g
        class="recharts-layer recharts-bar"
      >
        <g
          class="recharts-layer recharts-bar-rectangles"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
          </g>
        </g>
      </g>
    </svg>
    <div
      class="recharts-tooltip-wrapper"
      role="dialog"
      style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <div
        class="recharts-default-tooltip"
        style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); white-space: nowrap; color: white; border-radius: 5px;"
      >
        <p
          class="recharts-tooltip-label"
          style="margin: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<HorizontalBarChart /> > renders chart with expected canvas 2`] = `
<div>
  <div
    class="recharts-wrapper"
    role="region"
    style="position: relative; cursor: default; width: 800px; height: 400px;"
  >
    <svg
      class="recharts-surface"
      height="400"
      style="width: 100%; height: 100%;"
      viewBox="0 0 800 400"
      width="800"
    >
      <title />
      <desc />
      <defs>
        <clippath
          id="recharts5-clip"
        >
          <rect
            height="360"
            width="690"
            x="105"
            y="5"
          />
        </clippath>
      </defs>
      <g
        class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
      >
        <line
          class="recharts-cartesian-axis-line"
          fill="none"
          height="30"
          orientation="bottom"
          stroke="#666"
          width="690"
          x="105"
          x1="105"
          x2="795"
          y="365"
          y1="365"
          y2="365"
        />
        <g
          class="recharts-cartesian-axis-ticks"
        >
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="105"
              x2="105"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="105"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="105"
              >
                0
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="277.5"
              x2="277.5"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="277.5"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="277.5"
              >
                35,000
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="450"
              x2="450"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="450"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="450"
              >
                70,000
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="622.5"
              x2="622.5"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="622.5"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="622.5"
              >
                105,000
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="795"
              x2="795"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="795"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="795"
              >
                140,000
              </tspan>
            </text>
          </g>
        </g>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
      >
        <line
          class="recharts-cartesian-axis-line"
          fill="none"
          height="360"
          orientation="left"
          stroke="#666"
          style="font-size: .8rem;"
          width="100"
          x="5"
          x1="105"
          x2="105"
          y="5"
          y1="5"
          y2="365"
        />
        <g
          class="recharts-cartesian-axis-ticks"
        >
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              style="font-size: .8rem;"
              width="100"
              x="5"
              x1="99"
              x2="105"
              y="5"
              y1="95"
              y2="95"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              style="font-size: .8rem;"
              text-anchor="end"
              width="100"
              x="97"
              y="95"
            >
              <tspan
                dy="0.355em"
                x="97"
              >
                one
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              style="font-size: .8rem;"
              width="100"
              x="5"
              x1="99"
              x2="105"
              y="5"
              y1="275"
              y2="275"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              style="font-size: .8rem;"
              text-anchor="end"
              width="100"
              x="97"
              y="275"
            >
              <tspan
                dy="0.355em"
                x="97"
              >
                two
              </tspan>
            </text>
          </g>
        </g>
      </g>
      <g
        class="recharts-cartesian-grid"
      >
        <g
          class="recharts-cartesian-grid-horizontal"
        >
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="95"
            y2="95"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="275"
            y2="275"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="5"
            y2="5"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="365"
            y2="365"
          />
        </g>
        <g
          class="recharts-cartesian-grid-vertical"
        >
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="105"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="277.5"
            x2="277.5"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="450"
            x2="450"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="622.5"
            x2="622.5"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="795"
            x2="795"
            y="5"
            y1="5"
            y2="365"
          />
        </g>
      </g>
      <g
        class="recharts-layer recharts-bar"
      >
        <g
          class="recharts-layer recharts-bar-rectangles"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
          </g>
        </g>
      </g>
      <g
        class="recharts-layer recharts-bar"
      >
        <g
          class="recharts-layer recharts-bar-rectangles"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
          </g>
        </g>
      </g>
    </svg>
    <div
      class="recharts-tooltip-wrapper"
      role="dialog"
      style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <div
        class="recharts-default-tooltip"
        style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); white-space: nowrap; color: white; border-radius: 5px;"
      >
        <p
          class="recharts-tooltip-label"
          style="margin: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<HorizontalBarChart /> > renders chart with expected canvas 3`] = `
<div>
  <div
    class="recharts-wrapper"
    role="region"
    style="position: relative; cursor: default; width: 800px; height: 400px;"
  >
    <svg
      class="recharts-surface"
      height="400"
      style="width: 100%; height: 100%;"
      viewBox="0 0 800 400"
      width="800"
    >
      <title />
      <desc />
      <defs>
        <clippath
          id="recharts8-clip"
        >
          <rect
            height="360"
            width="690"
            x="105"
            y="5"
          />
        </clippath>
      </defs>
      <g
        class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
      >
        <line
          class="recharts-cartesian-axis-line"
          fill="none"
          height="30"
          orientation="bottom"
          stroke="#666"
          width="690"
          x="105"
          x1="105"
          x2="795"
          y="365"
          y1="365"
          y2="365"
        />
        <g
          class="recharts-cartesian-axis-ticks"
        >
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="105"
              x2="105"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="105"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="105"
              >
                0
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="277.5"
              x2="277.5"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="277.5"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="277.5"
              >
                35,000
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="450"
              x2="450"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="450"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="450"
              >
                70,000
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="622.5"
              x2="622.5"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="622.5"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="622.5"
              >
                105,000
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="690"
              x="105"
              x1="795"
              x2="795"
              y="365"
              y1="371"
              y2="365"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="30"
              orientation="bottom"
              stroke="none"
              text-anchor="middle"
              width="690"
              x="795"
              y="373"
            >
              <tspan
                dy="0.71em"
                x="795"
              >
                140,000
              </tspan>
            </text>
          </g>
        </g>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
      >
        <line
          class="recharts-cartesian-axis-line"
          fill="none"
          height="360"
          orientation="left"
          stroke="#666"
          style="font-size: .8rem;"
          width="100"
          x="5"
          x1="105"
          x2="105"
          y="5"
          y1="5"
          y2="365"
        />
        <g
          class="recharts-cartesian-axis-ticks"
        >
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              style="font-size: .8rem;"
              width="100"
              x="5"
              x1="99"
              x2="105"
              y="5"
              y1="65"
              y2="65"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              style="font-size: .8rem;"
              text-anchor="end"
              width="100"
              x="97"
              y="65"
            >
              <tspan
                dy="0.355em"
                x="97"
              >
                one
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              style="font-size: .8rem;"
              width="100"
              x="5"
              x1="99"
              x2="105"
              y="5"
              y1="185"
              y2="185"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              style="font-size: .8rem;"
              text-anchor="end"
              width="100"
              x="97"
              y="185"
            >
              <tspan
                dy="0.355em"
                x="97"
              >
                two
              </tspan>
            </text>
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis-tick"
          >
            <line
              class="recharts-cartesian-axis-tick-line"
              fill="none"
              height="360"
              orientation="left"
              stroke="#666"
              style="font-size: .8rem;"
              width="100"
              x="5"
              x1="99"
              x2="105"
              y="5"
              y1="305"
              y2="305"
            />
            <text
              class="recharts-text recharts-cartesian-axis-tick-value"
              fill="#666"
              height="360"
              orientation="left"
              stroke="none"
              style="font-size: .8rem;"
              text-anchor="end"
              width="100"
              x="97"
              y="305"
            >
              <tspan
                dy="0.355em"
                x="97"
              >
                max
              </tspan>
            </text>
          </g>
        </g>
      </g>
      <g
        class="recharts-cartesian-grid"
      >
        <g
          class="recharts-cartesian-grid-horizontal"
        >
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="65"
            y2="65"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="185"
            y2="185"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="305"
            y2="305"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="5"
            y2="5"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="795"
            y="5"
            y1="365"
            y2="365"
          />
        </g>
        <g
          class="recharts-cartesian-grid-vertical"
        >
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="105"
            x2="105"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="277.5"
            x2="277.5"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="450"
            x2="450"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="622.5"
            x2="622.5"
            y="5"
            y1="5"
            y2="365"
          />
          <line
            fill="none"
            height="360"
            stroke="#ccc"
            stroke-opacity="0.9"
            width="690"
            x="105"
            x1="795"
            x2="795"
            y="5"
            y1="5"
            y2="365"
          />
        </g>
      </g>
      <g
        class="recharts-layer recharts-bar"
      >
        <g
          class="recharts-layer recharts-bar-rectangles"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
          </g>
        </g>
      </g>
      <g
        class="recharts-layer recharts-bar"
      >
        <g
          class="recharts-layer recharts-bar-rectangles"
        >
          <g
            class="recharts-layer"
          >
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
            <g
              class="recharts-layer recharts-bar-rectangle"
            />
          </g>
        </g>
      </g>
    </svg>
    <div
      class="recharts-tooltip-wrapper"
      role="dialog"
      style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <div
        class="recharts-default-tooltip"
        style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); white-space: nowrap; color: white; border-radius: 5px;"
      >
        <p
          class="recharts-tooltip-label"
          style="margin: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;
